ssl/ssl_rsa.c

```diff
@@ -230,15 +232,17 @@ int SSL_CTX_use_certificate(SSL_CTX *ctx, X509 *x)
 static int ssl_set_cert(CERT *c, X509 *x)
 {
     EVP_PKEY *pkey;
-    size_t i;
+    size_t i=SSL_aANY; /* indicating non-legacy PKEY */
 
     pkey = X509_get0_pubkey(x);
     if (pkey == NULL) {
         ERR_raise(ERR_LIB_SSL, SSL_R_X509_LIB);
         return 0;
     }
 
-    if (ssl_cert_lookup_by_pkey(pkey, &i) == NULL) {
+    /* cert lookup test only for legacy keys */
+    if ((EVP_PKEY_get_base_id(pkey) != NID_undef) &&
+        (ssl_cert_lookup_by_pkey(pkey, &i) == NULL)) {
```

Comment:
This change is not correct. `i` here is an index into the pkeys array:

https://github.com/openssl/openssl/blob/dc0ef292f7df4ce0c49c64b47726a6768f9ac044/ssl/ssl_local.h#L2018

Each index corresponds to a difference certificate type:

https://github.com/openssl/openssl/blob/dc0ef292f7df4ce0c49c64b47726a6768f9ac044/ssl/ssl_local.h#L399-L408

You have initialised `i` to `SSL_aANY` and are using that value as a default in the case where there is no base id. But `SSL_aANY` isn't intended to be used as an index into pkeys. It's one of the auth types `SSL_a*`. These auth values are currently in the range 0 to 0x80.  As it happens `SSL_aANY` co-incidentally has a value of 0, which just happens to be a valid index corresponding to `SSL_PKEY_RSA`.

So this code will work - but only by accident because it pretends that the cert is actually an RSA cert.

Reply:
> But SSL_aANY isn't intended to be used as an index into pkeys. 

Got it . Becomes clear when running the code, too. This requires many more changes, incl. SIGALG_LOOKUP logic (and table) and probably many more. Sorry for the premature PR.